engine/src/main/java/org/pentaho/di/trans/steps/terafast/TeraFast.java

```diff
@@ -0,0 +1,475 @@
+              dataFilePrintStream.print( "Y" );
+            } else {
+              dataFilePrintStream.print( "N" );
+            }
+            break;
+          case ValueMetaInterface.TYPE_BINARY:
+            byte[] byt = rowMetaInterface.getBinary( row, i );
+            // REVIEW - this does an implicit byt.toString, which can't be what was intended.
+            dataFilePrintStream.print( byt );
+            break;
+          default:
+            throw new KettleException( BaseMessages.getString(
+              PKG, "TeraFast.Exception.TypeNotSupported", valueMeta.getType() ) );
+        }
+      }
+      dataFilePrintStream.print( FastloadControlBuilder.DATAFILE_COLUMN_SEPERATOR );
+    }
+    dataFilePrintStream.print( Const.CR );
+  }
+
+  private String pad( ValueMetaInterface valueMetaInterface, String data ) {
+    StringBuilder padding = new StringBuilder( data );
+    int padLength = valueMetaInterface.getLength() - data.length();
+    int currentPadLength = 0;
+    while ( currentPadLength < padLength ) {
+      padding.append( " " );
+      currentPadLength++;
+
+    }
+    return padding.toString();
+  }
+
+  /**
+   * Execute fastload.
+   *
+   * @throws KettleException
+   *           ...
+   */
+  public void execute() throws KettleException {
+    if ( this.meta.getTruncateTable().getValue() ) {
+      Database db = new Database( this, this.meta.getDbMeta() );
+      db.connect();
+      db.truncateTable( this.meta.getTargetTable().getValue() );
+      db.commit();
+      db.disconnect();
+    }
+    startFastLoad();
+
+    if ( this.meta.getUseControlFile().getValue() ) {
+      this.invokeLoadingControlFile();
+    } else {
+      this.invokeLoadingCommand();
+    }
+  }
+
+  /**
+   * Start fastload command line tool and initialize streams.
+   *
+   * @throws KettleException
+   *           ...
+   */
+  private void startFastLoad() throws KettleException {
+    final String command = this.createCommandLine();
+    this.logBasic( "About to execute: " + command );
+    try {
+      this.process = Runtime.getRuntime().exec( command );
+      new Thread( new ConfigurableStreamLogger(
+        getLogChannel(), this.process.getErrorStream(), LogLevel.ERROR, "ERROR" ) ).start();
+      new Thread( new ConfigurableStreamLogger(
+        getLogChannel(), this.process.getInputStream(), LogLevel.DETAILED, "OUTPUT" ) ).start();
+      this.fastload = this.process.getOutputStream();
+    } catch ( Exception e ) {
+      throw new KettleException( "Error while setup: " + command, e );
+    }
+  }
+
+  /**
+   * Invoke loading with control file.
+   *
+   * @throws KettleException
+   *           ...
+   */
+  private void invokeLoadingControlFile() throws KettleException {
+    File controlFile = null;
+    final InputStream control;
+    final String controlContent;
+    try {
+      controlFile = new File( resolveFileName( this.meta.getControlFile().getValue() ) );
+      control = FileUtils.openInputStream( controlFile );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *control*. The resource is closed at line 383. There are other execution paths that do not contain closure statements, for example, when FileUtils.readFileToString throws an exception. Ensure *control* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).